feat: superfluid typescript action provider

gtspencer opened the pull request:
[Link to Issue](https://github.com/coinbase/agentkit/issues/280)

## Description
Superfluid action providers, broken up by functionality (create stream, create pool, update stream, update pool). We experienced need for Superfluid functionality in our agent and decided to include the implementation in the core repo.

## Qualified Impact
Very little qualified impact due to the self contained nature of the action. Only dependency is the graphql-request package, which can be omitted in the event of an error.

#### Pool Actions:
- Create pool
- Update member units

#### Stream Actions:
- Create stream
- Update stream
- Close stream

#### Query Actions (easily extensible for other queries):
- Query open streams

#### Create Super Token
- Creates a SuperToken (wrapper) implementation for an underlying ERC20

#### Wrap Token
- Wraps an ERC20 into its Super Token implementation

## Tests

```
Chatbot: langchain-cdp-chatbot
Network: Base Mainnet
Setup: Add relevant actions (SuperfluidStreamActionProvider, etc.) to the chatbot
```

##### Create Super Token
```
Prompt: Can you create a super token for this erc20: 0xe8E55A847Bb446d967ef92F4580162fb8f2d3F38

-------------------
0x2D92C37A341f9C7AE0438179ee2835B9Bb692e40
Created super token for 0xe8E55A847Bb446d967ef92F4580162fb8f2d3F38.  Super token address at 0x2D92C37A341f9C7AE0438179ee2835B9Bb692e40  Transaction hash: 0xa9ea729bcbe30e3bb8610f9adf91fc4f4162e4cda5c33273c1f0d830388492da
-------------------
The super token has been successfully created for the ERC20 address `0xe8E55A847Bb446d967ef92F4580162fb8f2d3F38`. 

- **Super token address**: `0x2D92C37A341f9C7AE0438179ee2835B9Bb692e40`
- **Transaction hash**: `0xa9ea729bcbe30e3bb8610f9adf91fc4f4162e4cda5c33273c1f0d830388492da`

Let me know if you need further assistance!
-------------------
```
##### Wrap Super Token
```
Prompt: Can you wrap 1000 of this erc20 0xe8E55A847Bb446d967ef92F4580162fb8f2d3F38 into this super token 0x2D92C37A341f9C7AE0438179ee2835B9Bb692e40 on base mainnet

-------------------
Wallet Details:
- Provider: cdp_evm_wallet_provider
- Address: 0x77841ee4399EFfc9b515c09cf239D784392F897F
- Network:
  * Protocol Family: evm
  * Network ID: base-mainnet
  * Chain ID: 8453
- Native Balance: 37507762138705 WEI
-------------------

-------------------
Wrapped 1000 of token 0xe8E55A847Bb446d967ef92F4580162fb8f2d3F38 as a SuperToken.  Transaction hash: 0xd073bb4a85cbc79dc55f9996c500d141a257ccc2e173cce8f09017ec53b7cce6
-------------------
Successfully wrapped 1000 of the ERC20 token (0xe8E55A847Bb446d967ef92F4580162fb8f2d3F38) into the Super token. 

Transaction hash: **0xd073bb4a85cbc79dc55f9996c500d141a257ccc2e173cce8f09017ec53b7cce6**. 

If you have any more requests or questions, feel free to ask!
-------------------
```
##### Create Stream
```
Prompt: Can you create a superfluid stream of the super token 0x2D92C37A341f9C7AE0438179ee2835B9Bb692e40 at a rate of 100 to address 0x8F4359D1C2166452b5e7a02742D6fe9ca5448FDe

-------------------
Created stream of token 0x2D92C37A341f9C7AE0438179ee2835B9Bb692e40 to 0x8F4359D1C2166452b5e7a02742D6fe9ca5448FDe at a rate of 100. The link to the stream is https://app.superfluid.finance/stream/base-mainnet/0x77841ee4399EFfc9b515c09cf239D784392F897F-0x8F4359D1C2166452b5e7a02742D6fe9ca5448FDe-0x2D92C37A341f9C7AE0438179ee2835B9Bb692e40.  Transaction hash: 0xe108156834b91a52ab431e92095e171c93800d02f590271c5364eac26561dda3
-------------------
The Superfluid stream has been successfully created! You are streaming the SuperToken (0x2D92C37A341f9C7AE0438179ee2835B9Bb692e40) at a rate of 100 to the address **0x8F4359D1C2166452b5e7a02742D6fe9ca5448FDe**.

Here is the transaction hash for your reference: **0xe108156834b91a52ab431e92095e171c93800d02f590271c5364eac26561dda3**.

You can view the stream details using the following link: [View Stream](https://app.superfluid.finance/stream/base-mainnet/0x77841ee4399EFfc9b515c09cf239D784392F897F-0x8F4359D1C2166452b5e7a02742D6fe9ca5448FDe-0x2D92C37A341f9C7AE0438179ee2835B9Bb692e40).
-------------------
```
##### Pools
```
Prompt: Can you create a pool of the token 0x2D92C37A341f9C7AE0438179ee2835B9Bb692e40

-------------------
Created pool of token 0x2D92C37A341f9C7AE0438179ee2835B9Bb692e40 at 0x8651059749637291C4175d539d4eA62340aFBcAD.  Transaction hash: 0x0b717b24504452ebf18a3f707b882f247a25cff1d7302d21f71916f21b9379a5
-------------------
The pool for the token **0x2D92C37A341f9C7AE0438179ee2835B9Bb692e40** has been created successfully. 

- **Pool Address:** 0x8651059749637291C4175d539d4eA62340aFBcAD
- **Transaction Hash:** [0x0b717b24504452ebf18a3f707b882f247a25cff1d7302d21f71916f21b9379a5](https://etherscan.io/tx/0x0b717b24504452ebf18a3f707b882f247a25cff1d7302d21f71916f21b9379a5)

If you need any further assistance, feel free to ask!
-------------------
```

## Checklist

A couple of things to include in your PR for completeness:

- [x] Added documentation to all relevant README.md files
- [x] Added a changelog entry

cb-heimdall commented:
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td><code>2</code> if repo is sensitive</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>
<!-- CODEOWNERS & Bar Raisers -->
<!-- Power Rangers -->
<!-- Links -->

phdargen commented:
Hi @gtspencer, thanks for the contribution! Looks pretty good, a few initial comments below before I will do some manual testing

phdargen commented on typescript/.changeset/slimy-garlics-drum.md:
```diff
@@ -0,0 +1,5 @@
+---
+"@coinbase/agentkit": minor
```
minor -> patch

phdargen commented on typescript/agentkit/src/action-providers/superfluid/index.ts:
```diff
@@ -0,0 +1,4 @@
+export * from "./schemas";
+export * from "./superfluidStreamActionProvider";
+export * from "./superfluidPoolActionProvider";
+export * from "./superfluidQueryActionProvider";
```
might be good to add a 'SuperfluidActionProvider' wrapper combining the 3 action providers for convenience

phdargen commented on typescript/agentkit/src/action-providers/superfluid/superfluidPoolActionProvider.ts:
```diff
@@ -0,0 +1,125 @@
+      ).args;
+
+      if (success) {
+        return `Created pool of token ${args.erc20TokenAddress} at ${poolAddress}`;
```
please add txHashs in all returns

phdargen commented on typescript/agentkit/src/action-providers/superfluid/schemas.ts:
```diff
@@ -0,0 +1,55 @@
+/**
+ * Empty input schema
+ */
+export const EmptySchema = z.object({}).strip().describe("Empty input schema");
```
remove, doesnt seem to be used

phdargen commented on typescript/agentkit/src/action-providers/superfluid/schemas.ts:
```diff
@@ -0,0 +1,55 @@
+ */
+export const SuperfluidDeleteStreamSchema = z
+  .object({
+    erc20TokenAddress: z.string().describe("The ERC20 token to start streaming"),
```
suggest to add  .regex(/^0x[a-fA-F0-9]{40}$/, "Invalid Ethereum address format") to all addresses

phdargen commented on typescript/agentkit/src/action-providers/superfluid/superfluidPoolActionProvider.ts:
```diff
@@ -0,0 +1,125 @@
+
+      return `Updated member units of pool ${args.poolAddress} for member ${args.recipientAddress}, with new member units ${args.units}`;
+    } catch (error) {
+      return `Error creating Superfluid pool: ${error}`;
```
creating -> updating

phdargen commented on typescript/agentkit/src/action-providers/superfluid/superfluidQueryActionProvider.ts:
```diff
@@ -0,0 +1,57 @@
+
+      return `Current outflows are ${JSON.stringify(activeOutflows)}`;
+    } catch (error) {
+      return `Error creating Superfluid pool: ${error}`;
```
creating -> querying

phdargen commented on typescript/agentkit/src/action-providers/superfluid/superfluidStreamActionProvider.ts:
```diff
@@ -0,0 +1,189 @@
+
+      return `Updated stream of token ${args.erc20TokenAddress} to ${args.recipientAddress} at a rate of ${args.flowRate}`;
+    } catch (error) {
+      return `Error creating Superfluid stream: ${error}`;
```
creating -> updating

phdargen commented:
Main concern is Superfluid wrapped tokens vs erc20.

It is my understanding that you cant stream a erc20 token directly but need to supply a superfluid wrapped token.
At the very least, this should clearly be stated in the README and action descriptions.

Ideally, the createStream action would first wrap a given erc20 token before calling createFlow or alternatively a new action added that does the wrapping.

gtspencer commented:
hey @phdargen thanks for the review!  just addressed all your concerns:
- fixed minor issues involving hash response, networks, changeset type, and verbiage
- added a superfluid wrapper and Super Token creator action.  This way, if a stream action fails, the agent knows it may need to create the super token first and will prompt the user with that action
- Signed single commit

Ready for a new review at your leisure!

phdargen commented on typescript/agentkit/src/action-providers/superfluid/constants.ts:
```diff
@@ -0,0 +1,2828 @@
+  },
+] as const;
+
+export const SuperTokenFactoryAddress = "0xe20B9a38E0c96F61d1bA6b42a61512D56Fea1Eb3" as const;
```
superTokenFactory address is different on base-sepolia (0x7447E94Dfe3d804a9f46Bf12838d467c912C8F6C), see https://docs.superfluid.org/docs/protocol/contract-addresses. Would be good if users could try the workflow on testnet

phdargen commented on typescript/agentkit/src/action-providers/superfluid/superfluidActionProvider.ts:
```diff
@@ -0,0 +1,6 @@
+export * from "./schemas";
+export * from "./superfluidStreamActionProvider";
+export * from "./superfluidPoolActionProvider";
+export * from "./superfluidQueryActionProvider";
+export * from "./superfluidWrapperActionProvider";
+export * from "./superfluidSuperTokenCreatorActionProvider";
```
Please move exports back to index.ts. 

Instead add the following here that aggregates all superfluid actions a for a single import in eg chatbot files:
```suggestion
import { ActionProvider } from "../actionProvider";
import { EvmWalletProvider } from "../../wallet-providers";
import { Network } from "../../network";
import { SuperfluidStreamActionProvider } from "./superfluidStreamActionProvider";
import { SuperfluidPoolActionProvider } from "./superfluidPoolActionProvider";
import { SuperfluidQueryActionProvider } from "./superfluidQueryActionProvider";
import { SuperfluidWrapperActionProvider } from "./superfluidWrapperActionProvider";
import { SuperfluidSuperTokenCreatorActionProvider } from "./superfluidSuperTokenCreatorActionProvider";

/**
 * SuperfluidActionProvider aggregates all Superfluid-related actions into a single provider.
 */
export class SuperfluidActionProvider extends ActionProvider<EvmWalletProvider> {
  constructor() {
    super("superfluid", [
      new SuperfluidStreamActionProvider(),
      new SuperfluidPoolActionProvider(),
      new SuperfluidQueryActionProvider(),
      new SuperfluidWrapperActionProvider(),
      new SuperfluidSuperTokenCreatorActionProvider(),
    ]);
  }

  /**
   * Supports Base mainnet and Base sepolia like the underlying Superfluid providers.
   */
  supportsNetwork = (network: Network) =>
    network.networkId === "base-mainnet" || network.networkId === "base-sepolia";
}

export const superfluidActionProvider = () => new SuperfluidActionProvider();
```

phdargen commented on typescript/agentkit/src/action-providers/superfluid/README.md:
```diff
@@ -0,0 +1,187 @@
+# Superfluid Action Provider
+
+This directory contains the implementation for Superfluid, broken up by functionality:
```
Please add link to docs: https://docs.superfluid.org/

phdargen commented:
Please also add an entry in agentkit/README (https://github.com/coinbase/agentkit/blob/main/typescript/agentkit/README.md), should follow alphabetic order

phdargen commented on typescript/agentkit/src/action-providers/superfluid/schemas.ts:
```diff
@@ -0,0 +1,107 @@
+    erc20TokenAddress: z
+      .string()
+      .regex(/^0x[a-fA-F0-9]{40}$/, "Invalid Ethereum address format")
+      .describe("The ERC20 token to start streaming"),
```
token -> Super token

phdargen commented on typescript/agentkit/src/action-providers/superfluid/superfluidSuperTokenCreatorActionProvider.ts:
```diff
@@ -0,0 +1,104 @@
+      const receipt = await walletProvider.waitForTransactionReceipt(createSuperTokenHash);
+
+      const superTokenAddress = extractCreatedSuperTokenAddressAbi(receipt);
+      console.log(superTokenAddress);
```
please remove all console.log

phdargen commented on typescript/agentkit/src/action-providers/superfluid/superfluidSuperTokenCreatorActionProvider.ts:
```diff
@@ -0,0 +1,104 @@
+
+      const receipt = await walletProvider.waitForTransactionReceipt(createSuperTokenHash);
+
+      const superTokenAddress = extractCreatedSuperTokenAddressAbi(receipt);
```
Note: this fails for CdpSmartWallets as receipts doesnt include logs. You can ignore this, will look into it

phdargen commented:
> hey @phdargen thanks for the review! just addressed all your concerns:
> 
> * fixed minor issues involving hash response, networks, changeset type, and verbiage
> * added a superfluid wrapper and Super Token creator action.  This way, if a stream action fails, the agent knows it may need to create the super token first and will prompt the user with that action
> * Signed single commit
> 
> Ready for a new review at your leisure!

Hi @gtspencer, thanks for the update! I tested it and it works well except for some confusion with the units. Please have a look at my follow-up comments, then we should be able to get this in

gtspencer commented:
Hey @phdargen , thanks for the review!  Updated everything, made some changes to the prompts, and added more information on flow rate.  Let me know if there's anything else outstanding.  Thank you!

phdargen commented:
> Hey @phdargen , thanks for the review! Updated everything, made some changes to the prompts, and added more information on flow rate. Let me know if there's anything else outstanding. Thank you!

Hi @gtspencer, looks good! Just one more fix to use the correct contract depending on network

gtspencer commented:
oops!  Just updated!  @phdargen
